### DISCUSSION TITLE
Running code on server startup without a custom server

### DISCUSSION THREAD

**Discussion by KevinShiCA** [vercel/next.js#30826] [Ideas]
## Premise

Right now the only way to run some custom code on server startup is to follow the custom server.js pattern as described in [the docs](https://nextjs.org/docs/advanced-features/custom-server). This is heavy handed and has many drawbacks, namely Next's built in server optimizations are lost and the server cannot be deployed to Vercel. The docs also suggest that the custom server is only for cases where Next's built-in router isn't powerful enough, which isn't the case here.

## Use Case

My use case is that I want to set up a single database (Firebase) listener for the server's lifetime that gets real-time updates for a value that can be accessed during `getServerSideProps`. If I were to just query this value inside of `getServerSideProps`, it would cost one query per page load - with a single listener it only costs one query per value change.

## Existing Workaround

One possible workaround is to use a singleton pattern, such that an exported constant in a module only gets instantiated once per server lifetime. This works fine in production, but quickly runs into weirdness with hot reloading in a dev environment. It feels more like a hack than an actual solution.

## Proposal

Let the developer provide a `_startup.js` file that exports a default function to be run on server creation. Similarly to a custom `server.js`, this file will not be run through webpack and must be understandable by the current node version.

## POC Branch

A working proof-of-concept branch can be found here: https://github.com/vercel/next.js/compare/canary...KevinShiCA:run_on_server_startup

Note that prettier seems to disagree with some of the existing formatting - the only actual change is the addition of the `runCustomStartup` method, which gets called during `Server.prepare()`. This will check for the existence of a `_startup.js` file and attempt to `require()` it if found. Error handling and unit tests not implemented yet. (This is my first time working with the codebase, open to any feedback of where this code should go).

**Comment by cohendvir** [comment]
Running into the exact same issues with `Next.js` and this is very missing basic behavior.
I wish we had you're pr merged :)

**Comment by leerob** [comment]
Wanted to mention this solution. I know it's not perfect, but it should unblock you from running code before the Next.js server starts without needing to eject out of the default server.

```
node --require ./server-preload.js ./node_modules/.bin/next start
#    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#    Load our preload before running `next start`
```

Source and credit: https://jake.tl/notes/2021-04-04-nextjs-preload-hack
